### PR TITLE
Ignore unknown settings

### PR DESCRIPTION
**Pull request by zeha**
### Short description

Add a new `ignore-unknown-settings` setting, similar to ssh's `IgnoreUnknown`. Anything listed in there does not abort config parsing, even if PowerDNS does not know the setting otherwise.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

**Comment by zeha** [comment]
Rebased.

**Comment by zeha** [comment]
Rebased.

File: pdns/arguments.cc [hunk 1]
```diff
@@ -163,40 +197,28 @@ string ArgvMap::configstring(bool running, bool full)
   else
     help="# Autogenerated configuration file template\n\n";
 
+  // Effects parsing, should come first.
```
Comment (Habbie) [inline pdns/arguments.cc:200]:
```suggestion
  // Affects parsing, should come first.
```

**Review by Habbie — APPROVED** [review]
One comment typo, otherwise awesome! Will need review from Otto too. Also, you have one conflict.

File: pdns/reczones.cc [hunk 2]
```diff
@@ -362,17 +362,16 @@ std::shared_ptr<SyncRes::domainmap_t> parseAuthAndForwards()
```
Comment (omoerbeek) [inline pdns/reczones.cc:362]: These changes do not seem related, better leave them out.

File: pdns/arguments.cc [hunk 3]
```diff
@@ -365,23 +387,39 @@ void ArgvMap::parseOne(const string &arg, const string &parseOnly, bool lax)
...
+        d_params[var] = val;
         d_cleared.insert(var);
       }
     }
-    else if(!lax)
-      throw ArgException("Trying to set unknown parameter '"+var+"'");
+    else
+    {
+      // unknown setting encountered. see if its on the ignore list before throwing.
+      vector<string> parts;
```
Comment (omoerbeek) [inline pdns/arguments.cc:408]: Why not a set? saves you a loop below

File: docs/settings.rst [hunk 4]
```diff
@@ -678,6 +678,19 @@ or slave settings.
 
 Run within a guardian process. See :ref:`running-guardian`.
 
+.. _setting-ignore-unknown-settings:
```
Comment (omoerbeek) [inline docs/settings.rst:681]: .. versionadded:: xyz missing

File: pdns/recursordist/docs/settings.rst [hunk 5]
```diff
@@ -720,6 +720,19 @@ If set, EDNS options in incoming queries are extracted and passed to the :func:`
 
 If set, the root-hints are read from this file. If unset, default root hints are used.
 
+.. _setting-ignore-unknown-settings:
+
+``ignore-unknown-settings``
+---------------------------
+
```
Comment (omoerbeek) [inline pdns/recursordist/docs/settings.rst:727]: .. versionadded:: xyz missing

File: pdns/arguments.cc (hunk 3 above)
Comment (zeha) [reply pdns/arguments.cc]: set does not actually work with stringtok:

```
./misc.hh:102:17: error: no member named 'push_back' in 'std::__1::set<std::__1::basic_string<char>,
      std::__1::less<std::__1::basic_string<char> >, std::__1::allocator<std::__1::basic_string<char> > >'
      container.push_back (in.substr(i));
      ~~~~~~~~~ ^
arguments.cc:409:7: note: in instantiation of function template specialization 'stringtok<std::__1::set<std::__1::basic_string<char>,
      std::__1::less<std::__1::basic_string<char> >, std::__1::allocator<std::__1::basic_string<char> > > >' requested here
      stringtok(parts, d_params["ignore-unknown-settings"], " ,\t\n\r");
      ^
```

Comment (omoerbeek) [reply pdns/arguments.cc]: Bah, I don't see an immediate solution to this, `vector` and `set` are a bit too different...
You still might want to use `std::find` for the searching though

File: docs/settings.rst (hunk 4 above)
Comment (zeha) [reply docs/settings.rst:684]: added

File: pdns/arguments.cc [hunk 6]
```diff
@@ -365,23 +387,39 @@ void ArgvMap::parseOne(const string &arg, const string &parseOnly, bool lax)
...
+      parts.clear();
```
[the remark this answers is not in the document]
Comment (zeha) [reply pdns/arguments.cc]: removed, thanks

File: pdns/arguments.cc (hunk 3 above)
Comment (zeha) [reply pdns/arguments.cc]: I've added a stringtok template variant for set.

File: pdns/reczones.cc (hunk 2 above)
Comment (zeha) [reply pdns/reczones.cc]: gone!

File: pdns/arguments.cc (hunk 3 above)
Comment (omoerbeek) [reply pdns/arguments.cc]: Since `vector` actually both has `push_back` and `insert`, will the first template be selected in that case?

Comment (Habbie) [reply pdns/arguments.cc]: I don't think the `insert` variant would compile if it picked `vector`?

Comment (zeha) [reply pdns/arguments.cc]: `vector`s `insert` shouldn't compile, and IIRC non-compiling templates are not picked if there are other variants?

I haven't seen any trait types that would be useful here.

Comment (omoerbeek) [reply pdns/arguments.cc]: I was reading the template chapters of the C++ reference, but I couldn't find a clear description of the case at hand between all the thousands of words. The build does not seem to agree now though, but I did not look in detail.

Comment (zeha) [reply pdns/arguments.cc]: Yeah, for some reason now `ld` also hates it locally. Which is weird, because that should never be a symbol problem!?
I'll remove it.
